File: 3519.count-numbers-with-non-decreasing-digits.cpp
#
# @lc app=leetcode id=3519 lang=cpp
#
# [3519] Count Numbers with Non-Decreasing Digits
#
# @lc code=start
class Solution {
public:
    int countNumbers(string l, string r, int b) {
        const int MOD = 1e9 + 7;
        
        // Convert decimal string to base b digits
        auto toBase = [&](string num) {
            vector<int> result;
            while (!num.empty() && num != "0") {
                int remainder = 0;
                string next = "";
                for (char c : num) {
                    int digit = remainder * 10 + (c - '0');
                    if (!next.empty() || digit / b > 0) {
                        next += char('0' + digit / b);
                    }
                    remainder = digit % b;
                }
                result.push_back(remainder);
                num = next;
            }
            if (result.empty()) result.push_back(0);
            reverse(result.begin(), result.end());
            return result;
        };
        
        // Digit DP to count numbers with non-decreasing digits
        auto countUpTo = [&](vector<int>& digits) -> long long {
            int n = digits.size();
            map<tuple<int, int, bool, bool>, long long> memo;
            
            function<long long(int, int, bool, bool)> dp = [&](int pos, int prev, bool tight, bool started) -> long long {
                if (pos == n) {
                    return started ? 1 : 0;
                }
                
                auto key = make_tuple(pos, prev, tight, started);
                if (memo.count(key)) return memo[key];
                
                int limit = tight ? digits[pos] : (b - 1);
                long long res = 0;
                
                for (int d = 0; d <= limit; d++) {
                    if (started && d < prev) continue;
                    
                    bool newStarted = started || (d > 0);
                    bool newTight = tight && (d == limit);
                    int newPrev = newStarted ? d : 0;
                    
                    res = (res + dp(pos + 1, newPrev, newTight, newStarted)) % MOD;
                }
                
                return memo[key] = res;
            };
            
            return dp(0, 0, true, false);
        };
        
        // Subtract 1 from string
        auto subtract1 = [](string num) -> string {
            int i = num.size() - 1;
            while (i >= 0 && num[i] == '0') {
                num[i] = '9';
                i--;
            }
            if (i >= 0) {
                num[i]--;
            }
            size_t start = num.find_first_not_of('0');
            if (start == string::npos) return "0";
            return num.substr(start);
        };
        
        // Count for r
        vector<int> rDigits = toBase(r);
        long long countR = countUpTo(rDigits);
        
        // Count for l-1
        string lMinus1 = subtract1(l);
        long long countL = 0;
        if (lMinus1 != "0") {
            vector<int> lDigits = toBase(lMinus1);
            countL = countUpTo(lDigits);
        }
        
        return (countR - countL + MOD) % MOD;
    }
};
# @lc code=end

File: 3519.count-numbers-with-non-decreasing-digits.py
#
# @lc app=leetcode id=3519 lang=python3
#
# [3519] Count Numbers with Non-Decreasing Digits
#

# @lc code=start
from functools import lru_cache

class Solution:
    def countNumbers(self, l: str, r: str, b: int) -> int:
        MOD = 10**9 + 7

        def get_digits(n: int, base: int) -> list[int]:
            if n == 0:
                return []
            digits = []
            while n > 0:
                digits.append(n % base)
                n //= base
            digits.reverse()
            return digits

        def count_up_to(s: str, base: int) -> int:
            n = int(s)
            D = get_digits(n, base)
            if not D:
                return 0
            L = len(D)
            @lru_cache(maxsize=None)
            def dp(pos: int, prev: int, tight: int, started: int) -> int:
                if pos == L:
                    return 1 if started == 1 else 0
                ans = 0
                upper = D[pos] if tight == 1 else base - 1
                for d in range(upper + 1):
                    ntight = 1 if tight == 1 and d == D[pos] else 0
                    nstarted = 1 if started == 1 or d > 0 else 0
                    if nstarted == 0:
                        ans = (ans + dp(pos + 1, -1, ntight, 0)) % MOD
                    else:
                        if prev != -1 and d < prev:
                            continue
                        nprev = d
                        ans = (ans + dp(pos + 1, nprev, ntight, 1)) % MOD
                return ans
            return dp(0, -1, 1, 0)

        res_r = count_up_to(r, b)
        l_num = int(l)
        if l_num == 0:
            res_l = 0
        else:
            res_l = count_up_to(str(l_num - 1), b)
        return (res_r - res_l + MOD) % MOD

# @lc code=end